apis/Google.Cloud.Diagnostics.AspNet/Google.Cloud.Diagnostics.AspNet.Snippets/ErrorReportingSnippets.cs

```diff
@@ -0,0 +1,186 @@
+        /// Only for automatic logging, in self hosted web apis <see cref="HttpContext.Current"/>
+        /// is <code>null</code>.
+        /// </summary>
+        internal static void VerifyHttpContextLogging(ErrorEvent errorEvent)
+        {
+            Assert.Equal(HttpMethod.Get.Method, errorEvent.Context.HttpRequest.Method);
+            Assert.False(string.IsNullOrWhiteSpace(errorEvent.Context.HttpRequest.Url));
+        }
+
+        internal static string GetMessage(string message, string id) => $"{message} - {id}";
+    }
+
+    internal class ErrorReportingSnippetsApplication : HttpApplication
+    {
+        public const string Service = "service-name";
+        public const string Version = "version-id";
+        public static readonly string ProjectId = Utils.GetProjectIdFromEnvironment();
+
+        public void Configuration(IAppBuilder app)
+        {
+            HttpConfiguration config = new HttpConfiguration();
+            Register(config);
+            app.UseWebApi(config);
+        }
+
+        // Sample: RegisterExceptionLoggerWebApi
+        private static void Register(HttpConfiguration config)
+        {
+            // Add a catch all for the uncaught exceptions.
+            // Replace ProjectId with your Google Cloud Project ID.
+            // Replace Service with a name or identifier for the service.
+            // Replace Version with a version for the service.
+            config.Services.Add(typeof(System.Web.Http.ExceptionHandling.IExceptionLogger),
+                ErrorReportingExceptionLogger.Create(ProjectId, Service, Version));
+
+            // Web API routes
+            config.MapHttpAttributeRoutes();
+
+            config.Routes.MapHttpRoute(
+                name: "DefaultApi",
+                routeTemplate: "api/{controller}/{action}/{id}",
+                defaults: new { id = RouteParameter.Optional }
+            );
+        }
+        // End sample
+    }
+
+    public class ErrorReportingController : ApiController
+    {
+        public const string Service = "service-name";
+        public const string Version = "version-id";
+        public static readonly string ProjectId = Utils.GetProjectIdFromEnvironment();
+
+        /// <summary>Throws an <see cref="Exception"/>.
+        /// Needed to test the RegisterExceptionLoggerWebApi snippet.</summary>
+        [HttpGet]
+        public string ThrowsException(string id)
+        {
+            string message = ErrorReportingSnippets.GetMessage(nameof(ThrowsException), id);
+            throw new Exception(message);
+        }
+
+        // Sample: LogExceptionsWebApi
+        /// <summary>
+        /// This method catches an exception thrown by another method and explicitly
+        /// logs that exception.
+        /// </summary>
+        [HttpGet]
+        public string CatchesAndLogsException(string id)
+        {
+            GoogleWebApiExceptionLogger exceptionLogger = GoogleWebApiExceptionLogger.Create(ProjectId, Service, Version);
```

Comment:
Is this something we'd expect to appear in a controller method? I'd have thought it would be a dependency.

Reply:
Yes it should be, but there's no out of the box dependency resolution in asp.net as there is for core.
I can move this line to the controller level and add a comment there saying "Create the logger or use a dependency injection mechanism of your choice". I don't think we should include dependency resolution code as part of the snippet, tbh.